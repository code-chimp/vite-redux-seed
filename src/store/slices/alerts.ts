import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { v4 as uuid } from 'uuid';
import AlertTypes from '../../@enums/AlertTypes';
import AlertType from '../../@types/AlertType';
import IAlert from '../../@interfaces/IAlert';
import { IStore } from '../';

const initialState: Array<IAlert> = [];

const createAlertPayload = (type: AlertType, alert: Partial<IAlert>): IAlert => {
  return {
    ...alert,
    id: uuid(),
    type,
  };
};

export const alerts = createSlice({
  name: 'alerts',
  initialState,
  reducers: {
    removeAlert: (state: Array<IAlert>, action: PayloadAction<string>) => {
      const index = state.findIndex(a => a.id === action.payload);
      if (index > -1) {
        state.splice(index, 1);
      }
    },
    addErrorAlert: {
      reducer(state: Array<IAlert>, action: PayloadAction<IAlert>) {
        state.push(action.payload);
      },
      prepare(alert: Partial<IAlert>): any {
        return { payload: createAlertPayload(AlertTypes.Error, alert) };
      },
    },
    addInfoAlert: {
      reducer(state: Array<IAlert>, action: PayloadAction<IAlert>) {
        state.push(action.payload);
      },
      prepare(alert: Partial<IAlert>): any {
        return { payload: createAlertPayload(AlertTypes.Info, alert) };
      },
    },
    addSuccessAlert: {
      reducer(state: Array<IAlert>, action: PayloadAction<IAlert>) {
        state.push(action.payload);
      },
      prepare(alert: Partial<IAlert>): any {
        return { payload: createAlertPayload(AlertTypes.Success, alert) };
      },
    },
    addWarningAlert: {
      reducer(state: Array<IAlert>, action: PayloadAction<IAlert>) {
        state.push(action.payload);
      },
      prepare(alert: Partial<IAlert>): any {
        return { payload: createAlertPayload(AlertTypes.Warning, alert) };
      },
    },
  },
});

export const { addErrorAlert, addInfoAlert, addSuccessAlert, addWarningAlert, removeAlert } =
  alerts.actions;

export const selectAlerts = (state: IStore) => state.alerts;

export default alerts.reducer;
