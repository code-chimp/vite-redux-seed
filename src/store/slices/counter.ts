import { createSlice } from '@reduxjs/toolkit';
import { IStore } from '../';

export interface ICounterSlice {
  value: number;
}

export const counter = createSlice({
  name: 'counter',
  initialState: {
    value: 0,
  },
  reducers: {
    increment: (state: ICounterSlice) => {
      state.value += 1;
    },
    decrement: (state: ICounterSlice) => {
      state.value -= 1;
    },
    incrementByAmount: (state: ICounterSlice, action: { type: string; payload: number }) => {
      state.value += action.payload;
    },
  },
});

/* Ignore selectors in test coverage report due to simplicity */
/* istanbul ignore next */
export const selectCount = (state: IStore) => state.counter.value;

export const { increment, decrement, incrementByAmount } = counter.actions;

export default counter.reducer;
