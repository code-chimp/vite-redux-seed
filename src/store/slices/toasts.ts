import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { v4 as uuid } from 'uuid';
import IToastMessage from '../../@interfaces/IToastMessage';
import ToastTypes from '../../@enums/ToastTypes';
import { IStore } from '../';

const initialState: Array<IToastMessage> = [];

function createToastMessagePayload(toast: Partial<IToastMessage>): IToastMessage {
  return {
    ...toast,
    id: uuid(),
  };
}

export const toasts = createSlice({
  name: 'toasts',
  initialState,
  reducers: {
    removeToastMessage: (state: Array<IToastMessage>, action: PayloadAction<string>) => {
      const index = state.findIndex(t => t.id === action.payload);
      if (index > -1) {
        state.splice(index, 1);
      }
    },
    addErrorToastMessage: {
      reducer(state: Array<IToastMessage>, action: PayloadAction<IToastMessage>) {
        state.push(action.payload);
      },
      prepare(text: string): any {
        return { payload: createToastMessagePayload({ type: ToastTypes.Error, text }) };
      },
    },
    addInfoToastMessage: {
      reducer(state: Array<IToastMessage>, action: PayloadAction<IToastMessage>) {
        state.push(action.payload);
      },
      prepare(text: string): any {
        return { payload: createToastMessagePayload({ type: ToastTypes.Info, text }) };
      },
    },
    addSuccessToastMessage: {
      reducer(state: Array<IToastMessage>, action: PayloadAction<IToastMessage>) {
        state.push(action.payload);
      },
      prepare(text: string): any {
        return { payload: createToastMessagePayload({ type: ToastTypes.Success, text }) };
      },
    },
    addWarningToastMessage: {
      reducer(state: Array<IToastMessage>, action: PayloadAction<IToastMessage>) {
        state.push(action.payload);
      },
      prepare(text: string): any {
        return { payload: createToastMessagePayload({ type: ToastTypes.Warning, text }) };
      },
    },
  },
});

export const {
  addErrorToastMessage,
  addInfoToastMessage,
  addSuccessToastMessage,
  addWarningToastMessage,
  removeToastMessage,
} = toasts.actions;

export const selectToasts = (state: IStore) => state.toasts;

export default toasts.reducer;
