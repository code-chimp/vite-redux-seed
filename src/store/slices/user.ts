import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
import AsyncStates from '../../@enums/AsyncStates';
import IUser from '../../@interfaces/IUser';
import AsyncStatus from '../../@types/AsyncStatus';
import { fetchUsers } from '../../services/user/UsersApi';
import { IStore } from '../';

export interface IUserSlice {
  current: IUser | null;
  users: Array<IUser>;
  status: AsyncStatus;
  error: string | null;
}

export const initialState: IUserSlice = {
  current: null,
  users: [],
  status: AsyncStates.Idle,
  error: null,
};

export const loadUsers = createAsyncThunk(
  'users/fetchUsers',
  async (arg: void, { rejectWithValue }) => {
    try {
      return await fetchUsers();
    } catch (e: any) {
      return rejectWithValue(e.message || 'Internal Server Error');
    }
  },
);

export const user = createSlice({
  name: 'user',
  initialState,
  reducers: {
    clearCurrentUser: (state: IUserSlice) => {
      state.current = null;
    },
    setCurrentUser: (state: IUserSlice, action: { type: string; payload: number }) => {
      const user = state.users.find(x => x.id === action.payload);

      if (user) {
        state.current = user;
        state.error = null;
      } else {
        state.current = null;
        state.error = 'user not found';
      }
    },
  },
  extraReducers: builder => {
    builder.addCase(loadUsers.pending, (state: IUserSlice) => {
      state.status = AsyncStates.Pending;
    });

    builder.addCase(
      loadUsers.fulfilled,
      (state: IUserSlice, { payload }: PayloadAction<Array<IUser> | undefined>) => {
        state.status = AsyncStates.Success;
        state.users = payload ? (payload as Array<IUser>) : [];
        // or you could make it an additive operation
        // state.users = [...state.users, ...(action.payload as Array<IUser>)];
      },
    );

    builder.addCase(loadUsers.rejected, (state: IUserSlice, action: any) => {
      state.status = AsyncStates.Fail;
      state.error = action.payload as string;
      state.users = [];
    });
  },
});

/* Ignore selectors in test coverage report due to simplicity */
/* istanbul ignore next */
export const selectCurrentUser = (state: IStore) => state.user.current;

/* istanbul ignore next */
export const selectUsers = (state: IStore) => state.user.users;

export const { setCurrentUser, clearCurrentUser } = user.actions;

export default user.reducer;
