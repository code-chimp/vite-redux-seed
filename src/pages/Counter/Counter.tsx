import React, { FC } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faMinus, faPlus } from '@fortawesome/free-solid-svg-icons';
import { decrement, increment, selectCount } from '../../store/slices/counter';
import { useAppDispatch, useAppSelector } from '../../helpers';

const Counter: FC = () => {
  const count = useAppSelector(selectCount);
  const dispatch = useAppDispatch();

  return (
    <div className="row">
      <div className="col-12">
        <button
          aria-label="Increment value"
          data-testid="add-button"
          className="btn btn-success"
          onClick={() => dispatch(increment())}>
          <FontAwesomeIcon icon={faPlus} />
        </button>
        <span className="mx-5 fw-bold fs-3" data-testid="output-display">
          {count}
        </span>
        <button
          aria-label="Decrement value"
          data-testid="subtract-button"
          className="btn btn-danger"
          onClick={() => dispatch(decrement())}>
          <FontAwesomeIcon icon={faMinus} />
        </button>
      </div>
    </div>
  );
};

export default Counter;
