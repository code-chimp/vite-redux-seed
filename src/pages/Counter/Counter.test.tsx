/* eslint-disable no-magic-numbers */
import React from 'react';
import { Provider } from 'react-redux';
import configureMockStore, { MockStore } from 'redux-mock-store';
import { screen, fireEvent, render } from '@testing-library/react';
import { counter } from '../../store/slices/counter';
import Counter from './Counter';

const mockStore = configureMockStore([]);
let store: MockStore;

function getComponent(initialState: number): () => DocumentFragment {
  store = mockStore({
    counter: { value: initialState },
  });

  const { asFragment } = render(
    <Provider store={store}>
      <Counter />
    </Provider>,
  );

  return asFragment;
}

describe('components / Counter', () => {
  it('should match the snapshot', () => {
    const asFragment = getComponent(0);

    expect(asFragment()).toMatchSnapshot();
  });

  // Testing: Does the component dispatch expected actions
  it('should dispatch an increment action when the add button is clicked', () => {
    getComponent(0);
    const expected = [counter.actions.increment.type];

    const addButton = screen.getByTestId('add-button');
    expect(addButton).toBeInTheDocument();

    fireEvent.click(addButton);

    const received = store.getActions().map(action => action.type);
    expect(received).toEqual(expected);
  });

  it('should dispatch a decrement action when the subtract button is clicked', () => {
    getComponent(0);
    const expected = [counter.actions.decrement.type];

    const subtractButton = screen.getByTestId('subtract-button');
    expect(subtractButton).toBeInTheDocument();

    fireEvent.click(subtractButton);

    const received = store.getActions().map(action => action.type);
    expect(received).toEqual(expected);
  });

  it('should accurately reflect observed state', () => {
    getComponent(15);
    const expected = '15';

    const outputDisplay = screen.getByTestId('output-display');
    expect(outputDisplay.textContent).toBe(expected);
  });
});
