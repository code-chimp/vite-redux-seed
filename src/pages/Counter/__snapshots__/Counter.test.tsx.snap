// Vitest Snapshot v1

exports[`components / Counter > should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="row"
  >
    <div
      class="col-12"
    >
      <button
        aria-label="Increment value"
        class="btn btn-success"
        data-testid="add-button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-plus "
          data-icon="plus"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 80c0-17.7-14.3-32-32-32s-32 14.3-32 32V224H48c-17.7 0-32 14.3-32 32s14.3 32 32 32H192V432c0 17.7 14.3 32 32 32s32-14.3 32-32V288H400c17.7 0 32-14.3 32-32s-14.3-32-32-32H256V80z"
            fill="currentColor"
          />
        </svg>
      </button>
      <span
        class="mx-5 fw-bold fs-3"
        data-testid="output-display"
      >
        0
      </span>
      <button
        aria-label="Decrement value"
        class="btn btn-danger"
        data-testid="subtract-button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-minus "
          data-icon="minus"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M432 256c0 17.7-14.3 32-32 32L48 288c-17.7 0-32-14.3-32-32s14.3-32 32-32l352 0c17.7 0 32 14.3 32 32z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
